### ISSUE TITLE
Better toml reading code

### ISSUE THREAD

**Issue by tjayrush** [Great-Hill-Corporation/trueblocks-core#583]
The code for reading TOML files is hand-written. It would be better if we used a dedicated project, but I don't want it to affect the 'coding style' I've already established in many places throughout the code.

We should look for a package that we can 'hide behind' the existing interface. In other words, the changes needed to incorporate the new code would only require changing the library code, not any of the `tools` or `app` code.

**Comment by wildmolasses** [comment]
I understand that you want a library that can hide behind the existing interface.

I want to walk through an example that uses the existing interface. I'm looking at the code in [etherlibs/node.cpp#L938](https://github.com/Great-Hill-Corporation/quickBlocks/blob/develop/src/libs/etherlib/node.cpp#L938)

```cpp
string_q blockCachePath(const string_q& _part) {

        static string_q blockCache;
        if (blockCache.empty()) {
            CToml toml(configPath("quickBlocks.toml"));
            string_q path = toml.getConfigStr("settings", "blockCachePath", "<NOT_SET>");
            // cout << path << "\n";
            if (path == "<NOT_SET>") {
                path = configPath("cache/");
                toml.setConfigStr("settings", "blockCachePath", path);
                toml.writeFile();
            }
            CFilename folder(path);
            if (!folderExists(folder.getFullPath()))
                establishFolder(folder.getFullPath());
            if (!folder.isValid()) {
                cerr << "Invalid path (" << folder.getFullPath() << ") in config file. Quitting...\n";
                exit(0);
            }
            blockCache = folder.getFullPath();
            if (!endsWith(blockCache, "/"))
                blockCache += "/";
        }
        return substitute((blockCache + _part), "//", "/");
    }
```

According to your directions, this code would not be modified. Instead, we keep the existing interface, and switch out the methods within CToml. This means that toml.h will not be modified. Instead, toml.cpp will contain the changes.

I'm looking at this library for the toml parsing: https://github.com/skystrife/cpptoml

**Comment by tjayrush** [comment]
As you know, I hate dependencies which is one of the reasons I don't want to change the interface. Also, I want to make sure the library we select is fully open source and allows us to modify it in any way we want with no restrictions or impositions, so we need to be careful about the license it has. Also, ultimately, I will want to throw away a high percentage of their code since we won't use it and carry code that we don't use is a sure way to make difficult to maintain code. So--summary:

1. The right license so we have complete freedom of use,
2. Able to be hidden behind an interface without too much trouble so we can optimize it
3. Small, so we can make it smaller without a huge burden so when it's in our code it costs next to nothing to maintain
4. Faster than existing
5. Passes all tests

The most likely problem you will encounter is going to be my current implementation's shortcomings. So, like the doubles printing code, my code does weird things but my outside code relys on it.

Have fun.

**Comment by wildmolasses** [comment]
> The most likely problem you will encounter is going to be my current implementation's shortcomings. So, like the doubles printing code, my code does weird things that my outside code relies on which will break when the Toml code works correctly.

😭

**Comment by tjayrush** [comment]
This is turning out to be significantly more complicated than it should be because of the inadequacies of the existing implementation. We used cpptoml to try to replace our toml implementation, and it worked for almost all the test cases other than any that involved toml table arrays.

In our case this means four things:  whenBlock specials:list, qblocks tools:list, blockScrape exclusions:list, and monitor configs watches:list.

All but blockScrape actually could have been straight strings of JSON (which is what the current implementation expects). The new toml code cpptoml handles these items as real toml table arrays, so translating between the two is fraught with problems. I tried to build function on the cpptoml code to getConfigJson, but that was abortive (although probably the bext way to proceed).

The blockScrape use for exclusions reads the array as a straight string and parses based on '|' separator.

I'm going to postpone this issue until the next release. Re-enable it by searching for NEW_TOML #define.
